Dispatch and Network OS Object types should use DefaultOSObjectRetainTraits

ddkilzer opened the pull request:
#### 19967f9979a4d04f866009dfa181cc0d8829f15a
<pre>
Dispatch and Network OS Object types should use DefaultOSObjectRetainTraits
&lt;<a href="https://bugs.webkit.org/show_bug.cgi?id=304581">https://bugs.webkit.org/show_bug.cgi?id=304581</a>&gt;
&lt;<a href="https://rdar.apple.com/166988644">rdar://166988644</a>&gt;

Reviewed by Chris Dumez.

Extract DispatchOSObject.h and NetworkOSObject.h from
TypeCastsOSObject.h, then add template specializations for
WTF::DefaultOSObjectRetainTraits&lt;&gt; that are only active when ARC is
disabled since the default implementation in OSObjectPtr.h may be used
when ARC is enabled.

All of the uncommented changes below are to include the new headers.

* Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorXPCConnection.h:
* Source/WTF/WTF.xcodeproj/project.pbxproj:
- Add DispatchOSObject.h and NetworkOSObject.h to the Xcode project.
* Source/WTF/wtf/CMakeLists.txt:
- Add DispatchOSObject.h and NetworkOSObject.h to CMake.
* Source/WTF/wtf/MemoryPressureHandler.h:
* Source/WTF/wtf/WorkQueue.h:
* Source/WTF/wtf/cocoa/VectorCocoa.h:
* Source/WTF/wtf/darwin/DispatchOSObject.h: Add.
- Extract type cast support for libdispatch objects from
  TypeCastsOSObject.h.
- Add template specialization for WTF::DefaultOSObjectRetainTraits&lt;&gt;.
* Source/WTF/wtf/darwin/NetworkOSObject.h: Add.
- Extract type cast support for Network.framework objects from
  TypeCastsOSObject.h.
- Add template specialization for WTF::DefaultOSObjectRetainTraits&lt;&gt;.
* Source/WTF/wtf/darwin/TypeCastsOSObject.h:
- Move libdispatch and Network.framework code into DispatchOSObject.h
  and NetworkOSObject.h, respectively.
* Source/WebCore/SaferCPPExpectations/ForwardDeclCheckerExpectations:
- Remove bridge/objc/objc_runtime.h since it is now fixed.
* Source/WebCore/platform/FileMonitor.h:
* Source/WebCore/platform/cocoa/NetworkExtensionContentFilter.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
* Source/WebCore/platform/graphics/avfoundation/objc/QueuedVideoOutput.mm:
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h:
* Source/WebCore/platform/mac/PowerObserverMac.h:
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.h:
* Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.cpp:
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/cache/NetworkCacheData.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCTCPSocketCocoa.mm:
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCUDPSocketCocoa.mm:
* Source/WebKit/Platform/IPC/Connection.h:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
* Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm:
* Source/WebKit/UIProcess/ios/WKWebGeolocationPolicyDeciderIOS.mm:
* Source/WebKit/UIProcess/mac/ServicesController.h:
* Source/WebKit/webpushd/_WKMockUserNotificationCenter.mm:
* Source/WebKitLegacy/mac/WebView/WebPreferences.mm:
* Tools/DumpRenderTree/TestRunner.h:
* Tools/TestWebKitAPI/NetworkConnection.h:
* Tools/TestWebKitAPI/Tests/WTF/darwin/OSObjectPtr.cpp:
* Tools/TestWebKitAPI/Tests/WTF/darwin/TypeCastsOSObjectCF.cpp:
* Tools/TestWebKitAPI/Tests/WTF/darwin/TypeCastsOSObjectCocoa.mm:
* Tools/TestWebKitAPI/mac/VirtualGamepad.h:
* Tools/WebKitTestRunner/TestController.h:

Canonical link: <a href="https://commits.webkit.org/304891@main">https://commits.webkit.org/304891@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0149c9d13db5af25b44e82cfa32b0c71a3d8e57a

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136716 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/9075 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/48003 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/144454 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/89689 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b11319ef-a49c-496d-ad00-4e9fe0530477) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138588 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9776 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8921 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104555 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/76333 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/b95e0a30-4707-4a7e-8bc6-b59d6dc1c306) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139661 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/7152 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122504 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/85394 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6796 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/4482 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/5036 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/128677 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/116116 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40691 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/147202 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/135202 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8759 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/41264 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112904 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8777 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/7372 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/113233 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6715 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118794 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/62911 "Built successfully") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/21090 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8807 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36848 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/167982 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/8528 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/72373 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43826 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8747 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8599 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WTF/wtf/darwin/DispatchOSObject.h:
```diff
@@ -0,0 +1,82 @@
+struct DefaultOSObjectRetainTraits<TypeName##_t, std::false_type> { \
+    static ALWAYS_INLINE void retain(TypeName##_t ptr) \
+    { \
+        dispatch_retain(ptr); \
```
When I did the patch for xpc_retain, I had actually looked at libdispatch objects too and opted NOT to add custom logic for them. Based on my research at the time, dispatch_retain/dispatch_release and os_retain/os_release are functionally equivalent so it did not seem worth the extra code. Did you find differently?

ddkilzer replied:
Initially, no, but I took a second look and `dispatch_retain()` and `dispatch_release()` in non-Objective-C sources do have custom code.

I can't find `nw_retain()` and `nw_release()` doing a quick search, but they also have outlined (exported) functions for non-Objective-C sources.

ddkilzer replied:
@cdumez My other thought is that if the library or framework provides a custom retain/release method, we should probably use them for consistency and just to be good API citizens.

Commit:
Dispatch and Network OS Object types should use DefaultOSObjectRetainTraits
<https://bugs.webkit.org/show_bug.cgi?id=304581>
<rdar://166988644>

Reviewed by Chris Dumez.

Extract DispatchOSObject.h and NetworkOSObject.h from
TypeCastsOSObject.h, then add template specializations for
WTF::DefaultOSObjectRetainTraits<> that are only active when ARC is
disabled since the default implementation in OSObjectPtr.h may be used
when ARC is enabled.

All of the uncommented changes below are to include the new headers.

* Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorXPCConnection.h:
* Source/WTF/WTF.xcodeproj/project.pbxproj:
- Add DispatchOSObject.h and NetworkOSObject.h to the Xcode project.
* Source/WTF/wtf/CMakeLists.txt:
- Add DispatchOSObject.h and NetworkOSObject.h to CMake.
* Source/WTF/wtf/MemoryPressureHandler.h:
* Source/WTF/wtf/WorkQueue.h:
* Source/WTF/wtf/cocoa/VectorCocoa.h:
* Source/WTF/wtf/darwin/DispatchOSObject.h: Add.
- Extract type cast support for libdispatch objects from
  TypeCastsOSObject.h.
- Add template specialization for WTF::DefaultOSObjectRetainTraits<>.
* Source/WTF/wtf/darwin/NetworkOSObject.h: Add.
- Extract type cast support for Network.framework objects from
  TypeCastsOSObject.h.
- Add template specialization for WTF::DefaultOSObjectRetainTraits<>.
* Source/WTF/wtf/darwin/TypeCastsOSObject.h:
- Move libdispatch and Network.framework code into DispatchOSObject.h
  and NetworkOSObject.h, respectively.
* Source/WebCore/SaferCPPExpectations/ForwardDeclCheckerExpectations:
- Remove bridge/objc/objc_runtime.h since it is now fixed.
* Source/WebCore/platform/FileMonitor.h:
* Source/WebCore/platform/cocoa/NetworkExtensionContentFilter.h:
* Source/WebCore/platform/graphics/avfoundation/objc/MediaPlayerPrivateAVFoundationObjC.mm:
* Source/WebCore/platform/graphics/avfoundation/objc/QueuedVideoOutput.mm:
* Source/WebCore/platform/graphics/cocoa/VideoMediaSampleRenderer.h:
* Source/WebCore/platform/mac/PowerObserverMac.h:
* Source/WebCore/platform/mediastream/mac/ScreenCaptureKitCaptureSource.h:
* Source/WebCore/platform/network/cf/DNSResolveQueueCFNet.cpp:
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/cache/NetworkCacheData.h:
* Source/WebKit/NetworkProcess/cocoa/NetworkSessionCocoa.mm:
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCTCPSocketCocoa.mm:
* Source/WebKit/NetworkProcess/webrtc/NetworkRTCUDPSocketCocoa.mm:
* Source/WebKit/Platform/IPC/Connection.h:
* Source/WebKit/UIProcess/Cocoa/WebProcessPoolCocoa.mm:
* Source/WebKit/UIProcess/XR/ios/WKARPresentationSession.mm:
* Source/WebKit/UIProcess/ios/WKWebGeolocationPolicyDeciderIOS.mm:
* Source/WebKit/UIProcess/mac/ServicesController.h:
* Source/WebKit/webpushd/_WKMockUserNotificationCenter.mm:
* Source/WebKitLegacy/mac/WebView/WebPreferences.mm:
* Tools/DumpRenderTree/TestRunner.h:
* Tools/TestWebKitAPI/NetworkConnection.h:
* Tools/TestWebKitAPI/Tests/WTF/darwin/OSObjectPtr.cpp:
* Tools/TestWebKitAPI/Tests/WTF/darwin/TypeCastsOSObjectCF.cpp:
* Tools/TestWebKitAPI/Tests/WTF/darwin/TypeCastsOSObjectCocoa.mm:
* Tools/TestWebKitAPI/mac/VirtualGamepad.h:
* Tools/WebKitTestRunner/TestController.h:

Canonical link: https://commits.webkit.org/304891@main
